### PR TITLE
ELF Loader

### PR DESCRIPTION
**Pull request by JianyuWang0623**
## Summary
1. ELF_LOADTO_LMA
    - Load all sections to LMA not VMA, so the startup code(e.g. start.S) need relocate .data section to the final address(VMA) and zero .bss section by self.
    - For example, SiFli and Actions: Background: Device with small sram, Bootloader run in sram and psram, need boot to Application, with memory overlap and without XIP. VMA of .data is in "psram" and LMA in "rom", if not enable  `ELF_LOADTO_LMA` , ELF loader will load the section to VMA (will fill bootloader itself). For more details, please see Testing.1
2. ARCH_USE_COPY_SECTION
   - This option enables architecture-specific memory copy for dynamic code loading. For example, Ambiq has MRAM regions for instruction which can't load by the memcpy directly.

## Impact
Disabled by default

## Testing
1. Enable `ELF_LOADTO_LMA` so that the ELF loader loads .data to LMA, and do not zero .bss. The sections(.data, .bss, .etc) will be initialized in __start())
- start.c / start.S
```c
void __start(void)
{
    /* Zero .bss section */
    for (ptr = _sbss; ptr < _ebss;)
    {   
      *ptr++ = 0;
    }

    /* Copy .data section */
    for (src = _rom_data, dest = _sdata; dest < _edata;)
    { 
      *dest++ = *src++;
    }
}
```
- ld script
```lds
MEMORY
{ 
  rom (rx)            : ORIGIN = 0x18100000, LENGTH = 0x080000
  sram (rwx)          : ORIGIN = 0x01000400, LENGTH = 0x20000
  psram (rwx)         : ORIGIN = 0x18180000, LENGTH = 0x380000
// ... ...
}

SECTIONS
{
    _rom_data = LOADADDR(.data);
    .data : ALIGN(4) {
        _sdata = ABSOLUTE(.);
       // ... ...
        _edata = ABSOLUTE(.);
    } > psram AT > rom
    .bss : ALIGN(4) {
        _sbss = ABSOLUTE(.);
       // ... ...
        _ebss = ABSOLUTE(.);
    } > psram
}
```
- Header Info
```
$ readelf -l -S nuttx.elf 
Section Headers:
  [Nr] Name              Type            Addr     Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            00000000 000000 000000 00      0   0  0
// ... ...
  [ 5] .data             PROGBITS        18180000 070000 00056c 00  WA  0   0  4
// ... ...
  [ 8] .bss              NOBITS          18180570 070570 002998 00  WA  0   0 16
// ... ...

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  EXIDX          0x05940c 0x1814940c 0x1814940c 0x00008 0x00008 R   0x4
// ... ...
  LOAD           0x070000 0x18180000 0x1814a324 0x0056c 0x0056c RW  0x10000
  LOAD           0x000570 0x18180570 0x18180570 0x00000 0x02998 RW  0x10000

 Section to Segment mapping:
  Segment Sections...
   00     .ARM.exidx 
// ... ...
   03     .data 
   04     .bss 
```
```
Sections:
Idx Name          Size      VMA       LMA       File off  Algn
  0 .text         00049409  18100000  18100000  00010000  2**3
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
// ... ...
  4 .data         0000056c  18180000  1814a324  00070000  2**2
                  CONTENTS, ALLOC, LOAD, DATA
// ... ...
  7 .bss          00002998  18180570  18180570  00070570  2**4
                  ALLOC
// ... ...
```

2. The `ARCH_USE_COPY_SECTION` enables architecture-specific memory copy for dynamic code loading.
```c
int up_copy_section(void *dest, const void *src, size_t n)
{
  // ... ...
  if ((uint32_t)dest >= ADDR_MRAM_START && (uint32_t)dest < ADDR_MRAM_END)
    {
      mram->write(mram, dest, n, src);
    }
  else
    {
      memcpy(dest, src, n);
    }
  // ... ...
}
```

### CODE REVIEW

File: binfmt/libelf/libelf_uninit.c [hunk 1]
```diff
@@ -94,6 +94,12 @@ int elf_freebuffers(FAR struct elf_loadinfo_s *loadinfo)
 {
   /* Release all working allocations  */
 
+  if (loadinfo->phdr)
+    {
+      kmm_free((FAR void *)loadinfo->phdr);
+      loadinfo->phdr      = NULL;
```
Comment (pkarashchenko) [inline binfmt/libelf/libelf_uninit.c:100]:
```suggestion
      loadinfo->phdr = NULL;
```

File: include/nuttx/arch.h [hunk 2]
```diff
@@ -772,6 +772,21 @@ void up_textheap_free(FAR void *p);
...
+ * Description:
+ *   Copy section from general temporary buffer(src) to special addr(dest).
+ *
+ * Returned Value:
+ *   Zero (OK) on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+
+#if defined(CONFIG_ARCH_USE_COPY_SECTION)
+int up_copy_section(void *dest, const void *src, size_t n);
```
Comment (pkarashchenko) [inline include/nuttx/arch.h:787]:
```suggestion
int up_copy_section(FAR void *dest, FAR const void *src, size_t n);
```

**Review by pkarashchenko — APPROVED** [review]

**Comment by acassis** [comment]
@JianyuWang0623 please include the Summary in the commit log message

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by masayuki2009** [comment]
@JianyuWang0623 

I notched that the hello app (ELF) does not work with spresense:wifi_smp now.

```
nsh> ls -l /mnt/sd0/bin
/mnt/sd0/bin:
 -rw-rw-rw-    3672 hello
nsh> hello
[   24.857210] [CPU0] arm_hardfault: Hard Fault escalation:
[   24.859621] [CPU0] arm_hardfault: PANIC!!! Hard Fault!:
[   24.864840] [CPU0] arm_hardfault: 	IRQ: 3 regs: 0x2d06c7d0
[   24.870302] [CPU0] arm_hardfault: 	BASEPRI: 000000e0 PRIMASK: 00000000 IPSR: 00000003 CONTROL: 00000000
[   24.879671] [CPU0] arm_hardfault: 	CFSR: 00020000 HFSR: 40000000 DFSR: 00000000 BFAR: e000ed38 AFSR: 00000000
[   24.889589] [CPU0] arm_hardfault: Hard Fault Reason:
[   24.894533] [CPU0] _assert: Current Version: NuttX  12.0.0 8a13da322d Sep  4 2023 14:00:15 arm
[   24.903108] [CPU0] _assert: Assertion failed panic: at file: :0 task(CPU0): hello 0
[   24.910768] [CPU0] up_dump_register: R0: 00000001 R1: 2d06c0c0 R2: 2d06c0c0  R3: 00000001
[   24.918916] [CPU0] up_dump_register: R4: 00000000 R5: 00000001 R6: 2d06c0c0  FP: 00000000
[   24.927064] [CPU0] up_dump_register: R8: 00000000 SB: 00000000 SL: 00000000 R11: 00000000
[   24.935242] [CPU0] up_dump_register: IP: 00000000 SP: 2d06c8a8 LR: 0d00e213  PC: 00000000
[   24.943390] [CPU0] up_dump_register: xPSR: 80000000 BASEPRI: 000000e0 CONTROL: 00000000
[   24.951386] [CPU0] up_dump_register: EXC_RETURN: ffffffed
[   24.956787] [CPU0] dump_stack: User Stack:
[   24.960846] [CPU0] dump_stack:   base: 0x2d06c0d0
[   24.965515] [CPU0] dump_stack:   size: 00002024
[   24.970032] [CPU0] dump_stack:     sp: 0x2d06c8a8
[   24.974762] [CPU0] dump_tasks:    PID GROUP   CPU PRI POLICY   TYPE    NPX STATE   EVENT      SIGMASK          STACKBASE  STACKSIZE      USED   FILLED    COMMAND
[   24.989135] [CPU0] dump_tasks:   ----   ---  ---- --- -------- ------- --- ------- ---------- ---------------- 0x2d05afc8      2048       632    30.8%    irq
[   25.003204] [CPU0] dump_task:       0     0     0   0 FIFO     Kthread N-- Assigned           0000000000000000 0x2d05f964      1016       604    59.4%    CPU0 IDLE
[   25.017821] [CPU0] dump_task:       1     1     1   0 FIFO     Kthread N-- Running            0000000000000000 0x2d0603b0      1016       232    22.8%    CPU1 IDLE
[   25.032439] [CPU0] dump_task:       2     2     1 224 RR       Kthread --- Waiting Semaphore  0000000000000000 0x2d063ae8      2008       304    15.1%    hpwork 0x2d05cd44
[   25.047698] [CPU0] dump_task:       3     3     0  60 RR       Kthread --- Waiting Semaphore  0000000000000000 0x2d0646d8      2008       696    34.6%    lpwork 0x2d05cd2c
[   25.062926] [CPU0] dump_task:       5     5     0 200 RR       Task    --- Waiting MQ empty   0000000000000000 0x2d066278       992       456    45.9%    cxd56_pm_task
[   25.077879] [CPU0] dump_task:       6     6     0 100 RR       Task    --- Waiting Semaphore  0000000000000000 0x2d06aff0      3040      1728    56.8%    spresense_main
[   25.092924] [CPU0] dump_task:       7     7     0 100 RR       Task    --- Running            0000000000000000 0x2d06c0d0      2024       232    11.4%    hello

```

**Comment by masayuki2009** [comment]
@JianyuWang0623 

sabre-6quad:elf has errors. 

```
/home/ishikawa/opensource/QEMU/qemu-7.2.4/build/qemu-system-arm -semihosting -M sabrelite -m 1024 -smp 4 -kernel nuttx -nographic
Initial memory usage: 6228
Registering romdisk at /dev/ram0

Memory Usage after romdisk_register:
  Before:     6228 After:     6284 Change:       56
Mounting ROMFS filesystem at target=/mnt/romfs with source=/dev/ram0

Memory Usage after mount:
  Before:     6284 After:     6916 Change:      632

****************************************************************************
* Executing errno
****************************************************************************


Memory Usage after exec:
  Before:     6916 After:     6916 Change:        0
ERROR: exec(errno) failed: Unknown error 2

Memory Usage after program execution:
  Before:     6916 After:     6916 Change:        0

****************************************************************************
* Executing hello
****************************************************************************


Memory Usage after exec:
  Before:     6916 After:     6916 Change:        0
ERROR: exec(hello) failed: Unknown error 2

Memory Usage after program execution:
  Before:     6916 After:     6916 Change:        0
```

**Comment by masayuki2009** [comment]
@JianyuWang0623 

rv-virt:knsh64 crashed too.

```
QEMU_PATH=/home/ishikawa/opensource/QEMU/qemu-7.2.4/build
+ /home/ishikawa/opensource/QEMU/qemu-7.2.4/build/qemu-system-riscv64 -semihosting -nographic -cpu rv64 -smp 8 -M virt,aclint=on -bios none -kernel nuttx
ABC[    0.006000] _assert: Current Version: NuttX  12.0.0 8a13da322d Sep  4 2023 14:31:15 risc-v
[    0.006000] _assert: Assertion failed : at file: init/nx_bringup.c:302 task: Idle_Task 0x800017fc
[    0.006000] up_dump_register: EPC: 0000000080001e5e
[    0.006000] up_dump_register: A0: 0000000080200f30 A1: 000000000000012e A2: 0000000000000000 A3: 000000000000007e
[    0.006000] up_dump_register: A4: 0000000080200b28 A5: 0000000000000002 A6: 0000000080206d18 A7: fffffffffffffff8
[    0.006000] up_dump_register: T0: 000000000000002e T1: 0000000000000007 T2: 00000000000001ff T3: 0000000080206aa0
[    0.006000] up_dump_register: T4: 0000000080206a98 T5: 0000000000000009 T6: 000000000000002a
[    0.006000] up_dump_register: S0: 0000000000000000 S1: 0000000000000064 S2: 0000000080200b28 S3: 0000000000000000
[    0.006000] up_dump_register: S4: 0000000080018318 S5: 8000000200046022 S6: 0000000080201878 S7: 000000000000012e
[    0.006000] up_dump_register: S8: 0000000000000000 S9: 0000000000000000 S10: 0000000000000000 S11: 0000000000000000
[    0.006000] up_dump_register: SP: 0000000080206a00 FP: 0000000000000000 TP: 0000000000000000 RA: 0000000080001e5e
[    0.006000] dump_stack: User Stack:
[    0.006000] dump_stack:   base: 0x80206010
[    0.006000] dump_stack:   size: 00003056
[    0.006000] dump_stack:     sp: 0x80206a00
[    0.006000] stack_dump: 0x80206a00: 00000000 00000000 0032f8d0 00000000 000081ff 00000000 464c457f 00010102
[    0.006000] stack_dump: 0x80206a20: 80200b28 00000000 80200f30 00000000 00000000 00000000 80018318 00000000
[    0.006000] stack_dump: 0x80206a40: 0000012e 00000000 7474754e 00000058 00400000 00340035 00000000 00000000
[    0.006000] stack_dump: 0x80206a60: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 2e323100
[    0.006000] stack_dump: 0x80206a80: 00302e30 00000000 fffffffc ffffffff 6138ff83 61643331 64323233 70655320
[    0.006000] stack_dump: 0x80206aa0: 20342020 33323032 3a343120 313a3133 80200035 00000000 802087f0 00000000
[    0.006000] stack_dump: 0x80206ac0: 80206c00 73697200 00762d63 00000000 802087f0 00000000 802087f0 00000000
[    0.006000] stack_dump: 0x80206ae0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    0.006000] stack_dump: 0x80206b00: 00000000 00000000 00000000 00000000 00000000 00000000 80200b18 00000000
[    0.006000] stack_dump: 0x80206b20: 8020188c 00000000 00000064 00000000 00000c00 00000000 80007486 00000000
[    0.006000] stack_dump: 0x80206b40: 00000c00 00000000 80001a7a 00000000 00000000 00000000 80016400 00000000
[    0.006000] stack_dump: 0x80206b60: 00000000 00000000 00000c00 00000000 80200b28 00000000 80201890 00000000
[    0.006000] stack_dump: 0x80206b80: 80200b28 00000000 800019d2 00000000 80400000 00000000 00400000 00000000
[    0.006000] stack_dump: 0x80206ba0: 00000000 00000000 00000000 00000000 00000000 00000000 87e00000 00000000
[    0.006000] stack_dump: 0x80206bc0: 00000000 00000000 800008c8 00000000 00000000 00000000 80000976 00000000
[    0.006000] stack_dump: 0x80206be0: 00000000 00000000 00000000 00000000 00000000 00000000 8000004c 00000000
[    0.006000] dump_tasks:    PID GROUP PRI POLICY   TYPE    NPX STATE   EVENT      SIGMASK          STACKBASE  STACKSIZE      USED   FILLED    COMMAND
[    0.006000] dump_tasks:   ----   --- --- -------- ------- --- ------- ---------- ---------------- 0x802002b0      2048      2040    99.6%!   irq
[    0.006000] dump_task:       0     0   0 FIFO     Kthread N-- Running            0000000000000000 0x80206010      3056      1344    43.9%    Idle_Task
[    0.006000] dump_task:       1     1 100 RR       Kthread --- Waiting Unlock     0000000000000000 0x8020a040      1984       528    26.6%    lpwork 0x802015f0
```

**Comment by xiaoxiang781216** [comment]
@JianyuWang0623 please fix this problem. I guess that the crash is due to the relocatable object doesn't have program header.

**Comment by pkarashchenko** [comment]
I'm thinking if it is possible to add qemu runs to CI, so we can test more things in a virtual environment.

**Comment by JianyuWang0623** [comment]
@pkarashchenko  Please see https://github.com/apache/nuttx/pull/10474. Would you mind helping to verify with the patch? 
Verified with sim:elf. (Tried to build rv-virt:knnsh64 and sabre-6quad:elf but failed with problem of toolchain.)

**Comment by xiaoxiang781216** [comment]
> I'm thinking if it is possible to add qemu runs to CI, so we can test more things in a virtual environment.

we already have four defconfig to run test in every ci:
./boards/risc-v/qemu-rv/rv-virt/configs/citest/run
./boards/risc-v/qemu-rv/rv-virt/configs/citest64/run
./boards/sim/sim/sim/configs/citest/run
./boards/arm/imx6/sabre-6quad/configs/citest/run
But we need add more test to improv the coverage.

**Pull request merged**